### DISCUSSION TITLE
Create SQLite database file

### DISCUSSION THREAD

**Discussion by Stefan99353** [SeaQL/sea-orm#283] [Q&A] [answered]
SQLx does have the option to create the database file if it does not exist.
Diesel automatically creates the file by default.

I haven't found any way to do this using sea-orm.

**Comment by billy1624** [comment]
Hey @Stefan99353, welcome! You can do it via [from_sqlx_sqlite_pool](https://docs.rs/sea-orm/0.3.1/sea_orm/struct.SqlxSqliteConnector.html#method.from_sqlx_sqlite_pool). Basically, you create an ordinary SQLx SQLIte connection then convert it into `sea_orm::DatabaseConnection`.

**Comment by Stefan99353** [comment]
Thanks, that works.
Is there a way to do it in a way to be database agnostic?

**Comment by aaronleopold** [comment, marked as answer]
One thing I've been doing is just appending the appropriate permission flags to the end of the path to the database file in the `Database::connect` call. Since it uses sqlx internally i figured it would work and it seems to have done the trick. For example:

```rust
let connection = Database::connect("/Users/me/.myapp/data.db?mode=rwc").await?;
```

Just make sure the path to wherever you are creating the file exists beforehand (or create it using `std::fs`)

**Comment by tyt2y3** [comment]
Thank you @aaronleopold for helping out!

**Comment by Smerity** [comment]
Is it worth documenting this or is it assumed knowledge? I know my way around a few other languages / databases but had trouble finding any examples that created the SQLite database by default until (happily!) stumbling upon this. I also looked toward the sqlx docs but didn't find this (though I didn't look deeply as I was uncertain where to focus).

Almost all existing documented examples use in-memory SQLite. I wanted to play around, hence reaching immediately for SQLite, but also wanted persistence. If this might be appreciated I can look at adding it to documentation somewhere in SeaORM.

If this seems like a good idea I'll fork the [Database & Async Runtime page](https://www.sea-ql.org/SeaORM/docs/install-and-config/database-and-async-runtime) and propose a very brief mention re: SQLite memory / file `DATABASE_URL` next to "See also: [SQLx docs](https://docs.rs/crate/sqlx/latest/features)".

(Perhaps it's better in the [Connection Pool page](https://www.sea-ql.org/SeaORM/docs/install-and-config/connection), hmm...)

**Reply by tyt2y3** [reply]
Yeah sure a small note inside a bracket will be cool

**Reply by Samyak2** [reply]
Thanks for adding the example! I was stuck here too.

Edit: I just realized you're *the* smerity, did not expect to see you here :o

**Reply by tyt2y3** [reply]
Me too! Who did :D

**Comment by malteneuss** [comment]
Just wanted to add that now we need to specify the protocol:
```rust
let connection = Database::connect("sqlite://data.db?mode=rwc").await?;
```
Otherwise we get an error:

```
thread 'main' panicked at 'Fail to parse database URL: RelativeUrlWithoutBase', /home/.../sea-orm-0.12.3/src/database/db_connection.rs
```
